### ISSUE TITLE
update embedding of model prediction also, in the case of sparse tensors

### ISSUE THREAD

**Issue by vainaijr** [pytorch/pytorch#31651]
when we use,
```
x = nn.Embedding(10, 5, sparse=True)
```
then, only the embeddings at indices that are passed get updated, for example,
```
y = x(torch.LongTensor([0])).sum()
y.backward()
for param in x.parameters():
  print(param.grad)
```
gives
```
tensor(indices=tensor([[0]]),
       values=tensor([[1., 1., 1., 1., 1.]]),
       size=(10, 5), nnz=1, layout=torch.sparse_coo)
```
but if we want to update embedding of words in our vocabulary that our model predicted, then there is no way to do that, or changing requires_grad for a word in this embedding is also not possible, that is something like,
```
x.index[1].requires_grad = True
```
cannot be done, and only embedding of indices that are passed as input to this sparse tensor get updated.
there is no way to update embedding of a word that is not passed as input when using sparse embedding, and if we set sparse=False, then it will update all the words.

**Comment by albanD** [comment]
But the gradients will always be 0 if you don't use these indices no?
What do you expect to get here?

**Comment by vainaijr** [comment]
I wrote one example here, https://discuss.pytorch.org/t/different-prediction-results-update-different-weights-in-the-model/64997/6 

If we want to update the representations of words that are not part of input, that is update some words based on what model predicted.

**Comment by albanD** [comment]
Just a small, note, your model's `forward` is not returning the softmaxed output.

How do you want to change it? What would be the gradient for it?

**Comment by vainaijr** [comment]
from what I think, it would depend on position of words in the embedding space, that is, add to the gradients of words not passed as input, gradient_for_input_word/relative_distance.

For example, if we have vocab like,
[‘I’, ‘ate’, ‘eat’, ‘eating’, ‘an’, ‘apple’, ‘park’, ‘sleep’]
Task is to predict last word
Input is
‘I ate an ___’
And let us assume our model predicts the word ‘park’
Then after loss.backward(), some gradient will be there for input words, but we know that there are some words in our vocabulary, that are not passed as input, and we would like to change their representation also after each prediction model makes, in this case, words ‘eat’, ‘eating’ are similar to the word ‘ate’, so we would want to change their representation also, the word ‘park’ was predicted incorrectly, so change its representation also.

One way to do this could be using position information, and adding to the gradient for words not passed as input, (gradient_input/relative_distance),
here, these positions could be like (model will learn these),
[‘I’: 100, ‘an’: 200, ‘ate’: 300, ‘eat’: 301, ‘eating’: 302, ‘park’: 400, ‘apple’: 500, ‘sleep’: 600]
and we add to gradient for word ‘eat’, gradient_ate/(301-300), while for ‘eating’, gradient_ate/(302-300). Only if there are words nearby input in the embedding space then add these values, otherwise no gradient, it stays sparse, so we do not update gradient for the word ‘sleep’, as (600-300) is big, that is ‘sleep’ and ‘ate’ are far away in the embedding space.

**Comment by albanD** [comment]
That sounds like something very task specific rather than a general embedding feature no?
Also the fact that you use sparse gradient does not change anything right? You can use a hook to add new entries in the gradients in both cases.
